src/v2/style-guide/index.md

```diff
@@ -0,0 +1,1805 @@
+
+[![goto from xkcd.com](https://imgs.xkcd.com/comics/goto.png)](https://xkcd.com/292/)
+
+
+
+## Priority A Rules: Essential (Error Prevention)
+
+
+
+### Multi-word component names <sup data-p="a">essential</sup>
+
+**Component names should always be multi-word, except for root `App` components.**
+
+This prevents conflicts with existing and future HTML elements, since all HTML elements are a single word.
+
+{% raw %}<div class="style-example example-bad">{% endraw %}
+#### Bad
+
+``` js
+Vue.component('todo', {
+  // ...
+})
+```
+
+``` js
+export default {
+  name: 'Todo',
+  // ...
+}
+```
+{% raw %}</div>{% endraw %}
+
+{% raw %}<div class="style-example example-good">{% endraw %}
+#### Good
+
+``` js
+Vue.component('todo-item', {
+  // ...
+})
+```
+
+``` js
+export default {
+  name: 'TodoItem',
+  // ...
+}
+```
+{% raw %}</div>{% endraw %}
+
+
+
+### Component data <sup data-p="a">essential</sup>
+
+**Component `data` must be a function.**
+
+When using the `data` property on a component (i.e. anywhere except on `new Vue`), the value must be a function that returns an object.
+
+{% raw %}
+<details>
+<summary>
+  <h4>Detailed Explanation</h4>
+</summary>
+{% endraw %}
+
+When the value of `data` is an object, it's shared across all instances of a component. Imagine, for example, a `TodoList` component with this data:
+
+``` js
+data: {
+  listTitle: '',
+  todos: []
+}
+```
+
+We might want to reuse this component, allowing users to maintain multiple lists (e.g. for shopping, wishlists, daily chores, etc). There's a problem though. Since every instance of the component references the same data object, changing the title of one list will also change the title of every other list. The same is true for adding/editing/deleting a todo.
+
+Instead, we want each component instance to only manage its own data. For that to happen,each instance must generate a unique data object. In JavaScript, this can be accomplished by returning the object in a function:
+
+``` js
+data: function () {
+  return {
+    listTitle: '',
+    todos: []
+  }
+}
+```
+{% raw %}</details>{% endraw %}
+
+{% raw %}<div class="style-example example-bad">{% endraw %}
+#### Bad
+
+``` js
+Vue.component('some-comp', {
+  data: {
+    foo: 'bar'
+  }
+})
+```
+
+``` js
+export default {
+  data: {
+    foo: 'bar'
+  }
+}
+```
+{% raw %}</div>{% endraw %}
+
+{% raw %}<div class="style-example example-good">{% endraw %}
+#### Good
+``` js
+Vue.component('some-comp', {
+  data: function () {
+    return {
+      foo: 'bar'
+    }
+  }
+})
+```
+
+``` js
+// In a .vue file
+export default {
+  data () {
+    return {
+      foo: 'bar'
+    }
+  }
+}
+```
+
+``` js
+// It's OK to use an object directly in a root
```

Comment:
When it's "OK" to use an object in a root instance, I'd argue that it's still better to use a function for consistency and one less thing to remember. So maybe move this out of the "Good" section, for example as a note?

Reply:
It'd be pretty controversial to discourage this, since we've demonstrated the pattern in half the pages of the docs for a long time. 😕

Reply:
We don't discourage it, right? We can say that in the note, e.g.

"Since there's only one root Vue instance, it's OK to use a `data` object directly – in fact, that's what the demos on this site are doing. However, if you want consistency blahblah…"

Does it make sense?

Reply:
I think removing it from good would imply that it's _not_ good, right? Why move it to a note, when this section is literally for demoing what's acceptable?